src/Infrastructure/LeanCode.Kratos/KratosWebHookHandlerBase.cs

```diff
@@ -0,0 +1,106 @@
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using LeanCode.Kratos.Model;
+using Microsoft.AspNetCore.Http;
+
+namespace LeanCode.Kratos;
+
+public sealed record class KratosWebHookHandlerConfig(string ApiKey);
+
+public abstract partial class KratosWebHookHandlerBase
+{
+    private readonly Serilog.ILogger logger = Serilog.Log.ForContext<KratosWebHookHandlerBase>();
+
+    private readonly KratosWebHookHandlerConfig config;
+
+    protected KratosWebHookHandlerBase(KratosWebHookHandlerConfig config)
+    {
+        this.config = config;
+    }
+
+    protected virtual string ApiKeyHeaderName => "X-Api-Key";
+
+    protected abstract Task HandleCoreAsync(HttpContext ctx);
+
+    [SuppressMessage(
+        "?",
+        "CA1031",
+        Justification = "Exception is logged, rethrowing it for ASP.NET to handle is unnecessary."
+    )]
+    public async Task HandleAsync(HttpContext ctx)
+    {
+        try
+        {
+            if (VerifyApiKey(ctx.Request.Headers[ApiKeyHeaderName].ToString(), config.ApiKey))
+            {
+                await HandleCoreAsync(ctx);
+            }
+            else
+            {
+                logger.Error("Invalid Api Key");
+                ctx.Response.StatusCode = 403;
+            }
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Failed to process webhook");
+            ctx.Response.StatusCode = 500;
+        }
+    }
+
+    private static bool VerifyApiKey(string left, string right)
+    {
+        var leftBytes = MemoryMarshal.AsBytes(left.AsSpan());
+        var rightBytes = MemoryMarshal.AsBytes(right.AsSpan());
+
+        return CryptographicOperations.FixedTimeEquals(leftBytes, rightBytes);
+    }
+
+    public static Task WriteIdentityResponseAsync(HttpContext ctx, Identity identity)
+    {
+        ctx.Response.StatusCode = 200;
+        return ctx.Response.WriteAsJsonAsync(
+            new(identity),
+            KratosWebHookHandlerBaseJsonSerializerContext.Default.IdentityResponseBody,
+            cancellationToken: ctx.RequestAborted
+        );
+    }
+
+    public static Task WriteErrorResponseAsync(HttpContext ctx, List<ErrorMessage> messages, int statusCode = 400)
+    {
+        ctx.Response.StatusCode = statusCode;
+        return ctx.Response.WriteAsJsonAsync(
+            new(messages),
+            KratosWebHookHandlerBaseJsonSerializerContext.Default.ErrorResponseBody,
+            cancellationToken: ctx.RequestAborted
+        );
+    }
+}
+
+#if NET8_0_OR_GREATER
+[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.SnakeCaseLower)]
+#endif
+[JsonSerializable(typeof(IdentityResponseBody))]
+[JsonSerializable(typeof(ErrorResponseBody))]
+public partial class KratosWebHookHandlerBaseJsonSerializerContext : JsonSerializerContext;
+
+#pragma warning disable CA2227 // Collection properties should be read only
```

Comment:
Can we `SuppressMessage` instead of going the pragma way? It's more readable.

Reply:
It really isn't, lots of noise just to get rid of a warning that doesn't make sense in this context.